Ranked address space support.

kspirov opened the pull request:
* Rank is a separate class, with basic part and UUID.
* New DataType PROPOSAL and write operations based on rank and data type only when a rank is given (for quorum).
* Using the new read/write locks in StreamLogFiles (deadlock protection and multiple readers)
* Single address from the global space for the compact operation
* Bugfixes relatred to closing of  streams and channels (more to come)

kspirov commented:
I think this covers almost everything we need on the server side.

3 TODO for the server and we are done:

React the review comments (if any)

To figure out why the tests that open the StreamLogFiles don't work on my Windows notebook - this is usually a sign there is some unclosed read input stream. I noticed this long time ago, but as I, didn't pay too much attention as I am not using Windows often. But if there is a forgotten stream, we must profile and find it.

For compact() to discuss what will be the criteria the log file is for compacting. Should we read all entries (slow operation) or we should mark externaly the file has duplicate global addresses. Everything depends on how often compact() will be called. Currently the entries with the same global address will be compacted only if the compact is started because of trims() which I am not sure is correct.

corfudb-performance commented:
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **262d852**.

 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #490 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/T0TNwxZv2al1VW8Y5GHUFaaYjL5TkIpo)

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java:
```diff
@@ -725,7 +742,23 @@ public void append(LogAddress logAddress, LogData entry) {
+                if (entry.getRank()==null) {
+                    throw new OverwriteException();
+                } else {
+                    try (MultiReadWriteLock.AutoCloseableLock ignored = segmentLocks.acquireWriteLock(fh.getSegment())) {
```
Since append is serially called, is this lock necessary? also not write locks are acquired in write record too.

kspirov replied:
agree

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogWithRankedAddressSpace.java:
```diff
@@ -0,0 +1,37 @@
+            return Optional.of(Boolean.FALSE);
+        }
+        int compare = newEntry.getRank().compareTo(oldEntry.getRank());
+        return compare==0? Optional.empty(): compare<0? Optional.of(Boolean.FALSE): Optional.of(Boolean.TRUE);
```
Why bother returning empty? if compare == 0, then just return false (i.e. can't append) since the value already exists.

So if you are trying to write the same entry you'll throw overwrite exception and when you try to write with a lower rank  you get a rank exception.

This code is coalescing two different cases, lower rank writes are being interpreted as overwrites and same entry overwrites are being ignored.

kspirov replied:
Just a habit to write idempotent code. But in the quorum case we don't need it, nobody will retry the message, OK.

Maithem commented on runtime/src/main/java/org/corfudb/protocols/wireprotocol/IMetadata.java:
```diff
@@ -1,19 +1,17 @@
-import lombok.Data;
-import lombok.Getter;
-import lombok.RequiredArgsConstructor;
+import lombok.*;
```
nit: fix expand star imports

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java:
```diff
@@ -725,7 +742,23 @@ public void append(LogAddress logAddress, LogData entry) {
+                                throw new OverwriteException();
+                            }
+                        } else {
+                            log.debug("Duplicate requests for {} ", logAddress);
```
Do you care about the case where you have the same ranks, but the data in the entry differs ? would checking that be a good way to find out if some bad happened?

kspirov replied:
See below.

Maithem commented on test/src/test/java/org/corfudb/infrastructure/LogUnitServerTest.java:
```diff
@@ -209,5 +208,93 @@ public void testVerifyWithNoVerifyLog() throws Exception {
+                .data(new LogData(DataType.DATA, b))
+                .build();
+        m.setGlobalAddress(ADDRESS_0);
+        // m.setStreams(Collections.singleton(CorfuRuntime.getStreamID("a")));
```
remove comment

Maithem commented on infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java:
```diff
@@ -725,7 +742,23 @@ public void append(LogAddress logAddress, LogData entry) {
+                        if (canAppend.isPresent()) {
+                            if (canAppend.get()) {
+                                AddressMetaData addressMetaData = writeRecord(fh, logAddress.address, entry);
+                                fh.getKnownAddresses().put(logAddress.address, addressMetaData);
```
After the put, you can then trim the address of the old entry.

kspirov replied:
@Maithem  you cannot, as the log address doesn't change!  But my code makes sure the trim() will still work, just we need to see how to change criteria when to be activated.

The trick is that now the code which prepares the entries to reconstruct after trimming is linked hash map by exactly this log address, not array list. So the duplicate will disappear naturally, once trim is activated.

kspirov commented on infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java:
```diff
@@ -14,22 +14,11 @@
-import java.util.Map;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
```
... and setup that Intellij, OK :)

kspirov commented on infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java:
```diff
@@ -124,7 +124,7 @@ public LogUnitServer(ServerContext serverContext) {
             File dir = new File(logdir);
             if (!dir.exists()) {
-                dir.mkdir();
+                dir.mkdirs();
```
intentionally :)

kspirov commented:
Done. Going to the bed again, tomorrow I will reply your email.  :)

corfudb-performance commented:
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **e24af34**.

 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #490 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/tZ26rhoa36Y43xMa8I5wmIa5RK0sH0f0)

kspirov commented:
@Maithem Ops, I need to limit the commit message line to 72 chars for my Ubuntu env too.

kspirov commented:
I think we are OK with this part, now.

corfudb-performance commented:
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **b344770**.

 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #490 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/uNDS39H1YcF5TpM6itxJPg9o37JW2VsJ)

kspirov commented:
Returning back to "in progress", some changes in SteamLogWithRankedAddressSpace will be necessary

kspirov commented:
The server should be able to return back the value during phase1. Working on this, ETA today

kspirov commented:
I made the changes discussed with Dalia, now the client can adopt values. I believe this is the last change that concerns the server side.

In the last commit I have also some code in the LogUnitClient, but it is minimal and its purpose is just to test whether we are able to send and receive the new packets end to end.

corfudb-performance commented:
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **85f1cba**.

 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #490 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/P7kQIJ1upr97vnn6jIjlp1DcET5VceGk)

kspirov commented on infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogWithRankedAddressSpace.java:
```diff
@@ -0,0 +1,49 @@
+            throw new DataOutrankedException();
+        }
+        if (compare>0) {
+            if (newEntry.getType()==DataType.PROPOSAL && oldEntry.getType() != DataType.PROPOSAL) {
```
Currently PROPOSAL means just persisted entry after phase 1 - with unspecified value yet, but with reserved rank. I can change this if it confuses the reader.

So on phase 2 we would send DATA. I remember you discussed there should be assertion, that phase 2 should not be sent without phase 1, so we could change the naming. For example:

PHASE1_PROPOSAL and PHASE2_PROPOSAL?

dahliamalkhi replied:
got it.

Maithem commented:
Can you create a new pull request for your latest commit? Everything (clean up, stream fix, R/W lock) up till "Merge branch 'master' into server-quorum-stuff" seems to be good to be checked in. This pull request is getting large and its hard to review. Ignore, this if you're not adding more code to this pull request.

kspirov commented:
@Maithem  - if Dahlia is OK, of course you can merge the previous changes, it won't do harm. I can extract the last as a separate request, NP.

corfudb-performance commented:
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **e1f0415**.

 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #490 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/ITy3ZMMMT7mWx0wfMpzZg1cUqWcmwpDY)

kspirov commented:
@dahliamalkhi , @Maithem  - sorry, but I finally rebased everything on the top of the current master. I believe we won't need any general changes not only on replica side, but in the communication messages too. 

@Maithem - it is almost the same to the version you reviewed, but we have a new exception, that can return the data to the client, and the method in StreamLogFilesWithRankedAddressSpace is changed again - it directly throws the exception. As a side effect, the behavior when you receive the same rank is again idempotent (you remember you wanted a different behavior), but I don't think it will be a big harm. Otherwise this method would need to return 3-rd type of exception, only for this case.

@dahliamalkhi  - I changed the data type from PROPOSAL to RANK_ONLY, so it is more understandable and fixed the typo you showed.

kspirov commented on test/src/test/java/org/corfudb/infrastructure/LogUnitServerTest.java:
```diff
@@ -209,5 +208,90 @@ public void testVerifyWithNoVerifyLog() throws Exception {
+
+        WriteRequest m2 = WriteRequest.builder()
+                .writeMode(WriteMode.NORMAL)
+                .data(new LogData(DataType.DATA, b))
```
Just reminder - we decided, that this code is correct, but I renamed the data type from PROPOSAL to RANK_ONLY for better readability.

dahliamalkhi commented on infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogWithRankedAddressSpace.java:
```diff
@@ -0,0 +1,49 @@
+     * @throws ValueAdoptedException if the new message is a proposal during the two phase recovery write and there is an existing
+     * data at this log address already.
+     */
+    default void assertAppendPermitted(LogAddress logAddress, LogData newEntry) throws DataOutrankedException, ValueAdoptedException {
```
I believe that this method must be invoked with a lock on the logAddress' segment. If so, then to make sure no one invokes it without a lock in the future, please (a) indicate this in the comment, and (b) use a naming convention such as "unsafe" suffix or the like.

dahliamalkhi commented on runtime/src/main/java/org/corfudb/runtime/clients/LogUnitClient.java:
```diff
@@ -192,6 +211,29 @@ private static Object handleReadDataCorruption(CorfuPayloadMsg<ReadResponse> msg
+     * @param rank           The rank of this write]
+     *
+     */
+    CompletableFuture<Boolean> writeProposal(long address, Set<UUID> streams, IMetadata.DataRank rank) {
```
I hate to bug you again about terminology, but I think that 'proposal' may be a confusing term. It is sometimes used for the second phase of quorum-consensus.

could we change to 'writePrepare' ?

kspirov replied:
In my last client version I changed the signature of this method too and it is now  public CompletableFuture<Boolean> writeEmptyData(long address, DataType type, Set<UUID> streams, IMetadata.DataRank rank)  - we also supply DataType. I will port this and update in the reveiw now (I am in the airport)

kspirov commented:
@dahliamalkhi  Done, please wait Travis, I might have forgotten something
